lib/rules/no-restricted-imports.js

```diff
@@ -234,20 +241,53 @@ module.exports = {
         /**
          * Report a restricted path specifically for patterns.
          * @param {node} node representing the restricted path reference
-         * @param {Object} group contains a Ignore instance for paths, and the customMessage to show if it fails
+         * @param {Object} group contains an Ignore instance for paths, the customMessage to show on failure,
+         * and any restricted import names that have been specified in the config
+         * @param {Map<string,Object[]>} importNames Map of import names that are being imported
          * @returns {void}
          * @private
          */
-        function reportPathForPatterns(node, group) {
+        function reportPathForPatterns(node, group, importNames) {
             const importSource = node.source.value.trim();
 
-            context.report({
-                node,
-                messageId: group.customMessage ? "patternWithCustomMessage" : "patterns",
-                data: {
-                    importSource,
-                    customMessage: group.customMessage
+            const customMessage = group.customMessage;
+            const restrictedImportNames = group.importNames || [];
+
+            /*
+             * If we are not restricting to any specific import names and just the pattern itself,
+             * report the error and move on
+             */
+            if (restrictedImportNames.length < 1) {
+                context.report({
+                    node,
+                    messageId: customMessage ? "patternWithCustomMessage" : "patterns",
+                    data: {
+                        importSource,
+                        customMessage
+                    }
+                });
+                return;
+            }
+
+            restrictedImportNames.forEach(importName => {
```

Comment:
`import *` should be treated in the same way as in `paths` with `importNames`.

```js
/* eslint no-restricted-imports: ["error", {
    paths: [{
        name: "mod1",
        importNames: ["foo"]
    }],
    patterns: [{
        group: ["mod2"],
        importNames: ["bar"]
    }]
}]*/

import * as a from "mod1"; // this is reported

import * as b from "mod2"; // this isn't, but it should be

```

We might add an option to change this behavior (https://github.com/eslint/eslint/issues/15895), but either way `patterns` should be consistent with `paths` in this regard.

Reply:
Oh, weird. I would not expect `import * as a from "mod1";` to report an error unless it was imported as the matching import name (`foo`), but I can certainly update it to match the current behavior and move that test.